core/src/main/java/com/linecorp/armeria/common/outlier/DefaultOutlierDetector.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2024 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.outlier;
+
+import java.time.Duration;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.EventCount;
+import com.linecorp.armeria.common.util.EventCounter;
+import com.linecorp.armeria.common.util.Ticker;
+
+final class DefaultOutlierDetector implements OutlierDetector {
+
+    private final EventCounter counter;
+    private final double failureRateThreshold;
+    private final long minimumRequestThreshold;
+    private volatile boolean isOutlier;
```

Comment:
So if it once becomes an outlier, the client will never use it unlike Envoy bring it back after rejection time. Is that correct?
https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier#ejection-algorithm

Reply:
The outlier algorithm of Envoy is designed at the host (endpoint) level. Hosts do not tend to change for a long time. So, after ejection_time, sending requests again like a circuit breaker is necessary.

To provide the same functionality as Envoy, we may create an `OutlierDetectingEndpointGroup` using `OutlierDetector`. 
- If a host is determined as an outlier, the host is excluded from healthy endpoints for `ejection_time`.
- After `ejection_time`, the host is in the `endpoints`, and a new `OutlierDetector` is created.
- `ejection_time` will be provided by `OutlierDetectingEndpointGroup`

In fact, I implemented a similar `EndpointGroup` internally for the LINE push server. It would be better to modify the implementation and add it to Armeria.

Reply:
Understood. 👍